src/renderers/webgl/annotations/LineAnnotationRenderer.ts

```diff
@@ -0,0 +1,180 @@
+import { LineAnnotation, Coordinates, LinePoints } from '../../../types'
+import { DEFAULT_LABEL_STYLE, MIN_ANNOTATION_ZOOM, MIN_STROKE_ZOOM } from '../../../utils/constants'
+import { angle, distance } from '../../../utils/api'
+import { midPoint } from '../../../utils/webgl'
+import { Renderer } from '..'
+import LineSegment from '../objects/line/LineSegment'
+import LineStrokes from '../objects/line/LineStrokes'
+import Text from '../objects/text/Text'
+
+export default class LineAnnotationRenderer {
+  annotation: LineAnnotation
+
+  private _points: LinePoints = [0, 0, 0, 0]
+  private length = 0
+  private theta = 0
+
+  private fill: LineSegment
+  private strokes?: LineStrokes
+  private text?: Text
+
+  constructor(
+    private renderer: Renderer,
+    annotation: LineAnnotation
+  ) {
+    this.renderer = renderer
+    this.annotation = annotation
+    this.fill = new LineSegment(renderer.annotationsContainer, annotation.style)
+    this.setPoints(annotation.points)
+  }
+
+  update(annotation: LineAnnotation) {
+    this.annotation = annotation
+
+    this.fill.update(annotation.style.color, annotation.style.width, annotation.style.opacity)
+
+    if (this.strokes) {
+      this.strokes.update(annotation.style.stroke)
+    }
+
+    if (this.text) {
+      if (annotation.content) {
+        this.text.update(annotation.content, annotation.style.text)
+      } else {
+        this.text = this.managers.text.delete(this.text)
+      }
+    }
+
+    this.setPoints(annotation.points)
+
+    return this
+  }
+
+  render() {
+    const isVisible = this.visible()
+
+    if (isVisible && this.annotation.content && !this.text) {
```

Comment:
other renderer classes initialize and teardown objects in the update method, and only move/mount/unmount in the renderer. my hunch is that that's a better pattern than initialing Text here, and LineStrokes below. might be missing something, though.

Reply:
In large graphs, there is a sizable effort put on the app to instantiate required objects in each object's renderer; in most of these renderers the resulting object is not visible so the load is unnecessary. There are cases where instantiating 5K+ `PIXI.Text` objects simultaneously actually leads to crashes. We gain a substantial boost in performance if we lazily create these objects until the moment they are first needed. 

It makes sense to me to align this logic with the render's animation frame because this is where we apply the object's position (inclusive of any interpolation) as well as addressing culling/visibility. This allows us to reduce the times we are required to call the `visible` method.

There is potentially an argument for deferring this lazy instantiation to the `RenderObject` themselves (i.e. `Text`, `LineStrokes`, etc.). However, this would still require creating an internal class instance for every object yet this is likely less of a lift than instantiating any `PIXI` instance.

Reply:
could rework `visible` to optionally accept arguments `x`, `y` similar to `EdgeRenderer` so this logic is moved to `update`